File: Austin-Choi/PGS_완주하지못한선수.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

string solution(vector<string> participant, vector<string> completion) {
    string answer = "";
    //키 값 중복을 허용하지 않는 unordered_map 자료구조를 활용함
    //이름이 키 값으로 들어가고 value에는 존재하는 숫자가 들어감 
    //참가자에 a b c b c가 들어오면
    //자료구조에는 (a,1) (b,2) (c,2)로 저장되게 될 것임.
    unordered_map<string, int> pMap;

    //참가자 이름을 키값으로 같은 이름이 나올때마다 +1을 해줌
    for (auto name : participant) {
        ++pMap[name];
    }

    //완주한 선수 이름을 키값으로 접근해 해당 이름의 갯수를 -1을 해줌
    for (auto name : completion) {
        --pMap[name];
    }

    //만약 키로 접근했을때 value값이 0이 아니라면 완주를 하지 못하고 
    //남아있다는 뜻이므로 value값이 0 보다 클 때 
    //키값을 반환함
    for (auto item : pMap) {
        if (item.second > 0) {
            answer = item.first;
            break;
        }
    }
    return answer;
}

File: 2oo1s/PGS_완주하지못한선수.java
import java.util.*;

class Solution {
	public String solution(String[] participant, String[] completion) {
		String answer = "";

		// 선수 이름을 key값, participant 배열에 선수 이름이 몇번 있는지를 value값으로 
		Map<String, Integer> runner = new HashMap<>();

		for (String name : participant) {
			if (runner.get(name) == null)
				runner.put(name, 1);
			// 이름이 1번 이상 나온 선수면 value값 +1 
			else
				runner.put(name, runner.get(name) + 1);
		}

		// completion에 있는 선수를 runner Map에서 찾아서 value값 -1 
		for (String name : completion) {
			runner.put(name, runner.get(name) - 1);
		}

		// value값이 0이 아닌 선수가 완주 못한 선수이므로 answer에 저장 
		for (String name : runner.keySet()) {
			if (runner.get(name) != 0) {
				answer = name;
				break;
			}
		}

		return answer;
	}
}